src/rdkafka_telemetry_encode.c

```diff
@@ -0,0 +1,869 @@
+                &calculate_queue_time_avg,
+            [RD_KAFKA_TELEMETRY_METRIC_PRODUCER_RECORD_QUEUE_TIME_MAX] =
+                &calculate_queue_time_max,
+};
+
+static const rd_kafka_telemetry_metric_value_calculator_t
+    CONSUMER_METRIC_VALUE_CALCULATORS[RD_KAFKA_TELEMETRY_CONSUMER_METRIC__CNT] = {
+        [RD_KAFKA_TELEMETRY_METRIC_CONSUMER_CONNECTION_CREATION_RATE] =
+            &calculate_connection_creation_rate,
+        [RD_KAFKA_TELEMETRY_METRIC_CONSUMER_CONNECTION_CREATION_TOTAL] =
+            &calculate_connection_creation_total,
+        [RD_KAFKA_TELEMETRY_METRIC_CONSUMER_NODE_REQUEST_LATENCY_AVG] =
+            &calculate_broker_avg_rtt,
+        [RD_KAFKA_TELEMETRY_METRIC_CONSUMER_NODE_REQUEST_LATENCY_MAX] =
+            &calculate_broker_max_rtt,
+        [RD_KAFKA_TELEMETRY_METRIC_CONSUMER_COORDINATOR_ASSIGNED_PARTITIONS] =
+            &calculate_consumer_assigned_partitions,
+};
+
+static const char *get_client_rack(const rd_kafka_t *rk) {
+        return rk->rk_conf.client_rack &&
+                       RD_KAFKAP_STR_LEN(rk->rk_conf.client_rack)
+                   ? (const char *)rk->rk_conf.client_rack->str
+                   : NULL;
+}
+
+static const char *get_group_id(const rd_kafka_t *rk) {
+        return rk->rk_conf.group_id_str ? (const char *)rk->rk_conf.group_id_str
+                                        : NULL;
+}
+
+static const char *get_group_instance_id(const rd_kafka_t *rk) {
+        return rk->rk_conf.group_instance_id
+                   ? (const char *)rk->rk_conf.group_instance_id
+                   : NULL;
+}
+
+static const char *get_member_id(const rd_kafka_t *rk) {
+        return rk->rk_cgrp->rkcg_member_id &&
+                       rk->rk_cgrp->rkcg_member_id->len > 0
+                   ? (const char *)rk->rk_cgrp->rkcg_member_id->str
+                   : NULL;
+}
+
+static const char *get_transactional_id(const rd_kafka_t *rk) {
+        return rk->rk_conf.eos.transactional_id
+                   ? (const char *)rk->rk_conf.eos.transactional_id
+                   : NULL;
+}
+
+static const rd_kafka_telemetry_attribute_config_t producer_attributes[] = {
+    {"client_rack", get_client_rack},
+    {"transactional_id", get_transactional_id},
+};
+
+static const rd_kafka_telemetry_attribute_config_t consumer_attributes[] = {
+    {"client_rack", get_client_rack},
+    {"group_id", get_group_id},
+    {"group_instance_id", get_group_instance_id},
+    {"member_id", get_member_id},
+};
+
+static int
+count_attributes(rd_kafka_t *rk,
+                 const rd_kafka_telemetry_attribute_config_t *configs,
+                 int config_count) {
+        int count = 0, i;
+        for (i = 0; i < config_count; ++i) {
+                if (configs[i].getValue(rk)) {
+                        count++;
+                }
+        }
+        return count;
+}
+
+static void set_attributes(rd_kafka_t *rk,
+                           rd_kafka_telemetry_resource_attribute_t **attributes,
```

Comment:
Here let's pass the array of structs instead of the pointer to that array.
```suggestion
                           rd_kafka_telemetry_resource_attribute_t *attributes,
```

Reply:
Thanks, fixed.